### ISSUE TITLE
Integrate CDI-like framework to sponge api.

### ISSUE THREAD

**Issue by NeumimTo** [SpongePowered/SpongeAPI#1156]
Web java (Spring or JEE) comes with neat automatic dependency injection. It would be nice to have similar approach in Sponge.

If is there anyone who has never worked with web java before lets have a simple example (Spring) how the system works.  (In the example i wont mention things such as `@Scope` and `@Qualifier`, `@Resource` ... it would be too long)

In spring we have 4 basic annotations

```
@Component  - stereotype for any managed component 
@Repository   -  stereotype for persistence layer          
@Controller     - stereotype for presentation layer          
@Service        - stereotype for service layer                

@Repository
public class MyEntityDAO {

          public void save(MyEntity ........
} 

@Service
public class MyEntityService {

     @Autowired
     private MyEntityDAO dao

     public void save(MyEntity entity) {dao.save(entity));}

}
@Configuration
@ComponentScan(basePackage = "my.package")
public void Configuration {}
```

In a nutshell spring scans all classes in a defined package, once it finds a class annotated with component annotation its instance is created and fields with autowired annotation are injected with a corresponding type.

Why should you bother about this?

```
 - Plugin development will become faster
 - Easier to hook one plugin into another.
 - Reduces code you have to write
 - If used correctly helps you separate service and persistance layers which leads to cleaner and more readable code
 - With feature like this you could also automatically register all listeners and commands executors, this could reduce stupid bugs eg.: when someone forgets to register a listener.
```

**Comment by Xfel** [comment]
Right now, sponge is using google guice to perform dependency injection. You can already have custom objects injected using guice's JIT-binding features. 

Unfortunately, those bindings don't always work the way it is expected as they are created in the shared parent injector, not the concrete plugin injector. This does allow cross plugin injection, but I think we should not do that as it might lead to unexpected result. We have the services for cross-plugin injection, and that is all. 
That said, a way to use dependency injection for services would really be helpful.

Another important point is configuration. Ideally, I could select ie. whether to use a file-tree or sql based implementation of my data storage before injection. However, that would require us to read the config before we create the injector, which would completely break the current lifecycle...

**Comment by NeumimTo** [comment]
guice requires you to write a lot boilerplate, i dont really feel comfortable to work with it.


> Another important point is configuration. Ideally, I could select ie. whether to use a file-tree or sql based implementation of my data storage before injection. However, that would require us to read the config before we create the injector, which would completely break the current lifecycle...

How exactly?

**Comment by Xfel** [comment]
Currently, the plugin injector is created before the plugin instance is created. We should read the configuration during the preinit phase, which will only happen after the plugin instances were created. so if we want to select an implementation based on some config value, we'd need to read the config earlier.

**Comment by ryantheleach** [comment]
@NeumimTo Unless you have specific suggestions on how this can apply to Minecraft plugins (and the SpongeAPI) specifically, rather then web requests / servers, I'm going to keep this ticket closed as in it's current form feels too abstract. There's nothing stopping someone (to my knowledge) creating a Proof of Concept plugin on how this should work.
